R/paginate.R

```diff
@@ -39,29 +47,55 @@
 #' request("https://pokeapi.co/api/v2/pokemon") %>%
 #'   req_url_query(limit = page_size) %>%
 #'   req_paginate_next_url(
-#'     parse_resp = resp_body_json,
-#'     next_url = function(resp, parsed) parsed[["next"]],
-#'     n_pages = function(resp, parsed) {
+#'     parse_resp = function(resp) {
+#'       parsed <- resp_body_json(resp)
+#'       results <- parsed$results
+#'       data <- data.frame(
+#'         name = sapply(results, `[[`, "name"),
+#'         url = sapply(results, `[[`, "url")
+#'       )
+#'
+#'       list(data = data, next_url = parsed$`next`)
+#'     },
+#'     n_pages = function(parsed) {
 #'       total <- parsed$count
 #'       ceiling(total / page_size)
 #'     }
 #'   )
 req_paginate <- function(req,
                          next_request,
                          parse_resp = NULL,
+                         required_fields = NULL,
                          n_pages = NULL) {
   check_request(req)
-  check_function2(next_request, args = c("req", "resp", "parsed"))
+  check_function2(next_request, args = c("req", "parsed"))
   check_function2(parse_resp, args = "resp", allow_null = TRUE)
-  parse_resp <- parse_resp %||% identity
-  check_function2(n_pages, args = c("resp", "parsed"), allow_null = TRUE)
-  n_pages <- n_pages %||% function(resp, parsed) Inf
+  parse_resp <- parse_resp %||% function(resp) list(data = resp)
+  check_character(required_fields, allow_null = TRUE)
+  required_fields <- union(required_fields, "data")
+  check_function2(n_pages, args = "parsed", allow_null = TRUE)
+  n_pages <- n_pages %||% function(parsed) Inf
+
+  wrapped_parse_resp <- function(resp) {
+    out <- parse_resp(resp)
+    if (!is.list(out)) {
```

Comment:
Since we're taking a dependency on vctrs, I think this could now be `vctrs::obj_check_list()`